File: backend/models/learning.py
"""Continual learning models — architecture §12."""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class TradeOutcomeLabel(str, Enum):
    win = "win"
    loss = "loss"
    scratch = "scratch"


class FailureMemoryMatch(BaseModel):
    """A past losing context similar to the current recommendation (§12.6)."""

    failure_id: str
    similarity: float
    summary: str
    loss_pnl_inr: float
    lesson: str
    strategy: str
    underlying_symbol: str
    closed_at: datetime


class LearningInsight(BaseModel):
    """Per-recommendation learning overlay shown in the insight packet."""

    failure_matches: list[FailureMemoryMatch] = Field(default_factory=list)
    confidence_penalty: float = 0.0
    confidence_before: float
    confidence_after: float
    module_win_rate: float | None = None
    module_trade_count: int = 0
    module_expectancy_inr: float | None = None
    learning_note: str


class OpenTradeRecord(BaseModel):
    """Open discretionary trade awaiting outcome for the learning loop."""

    trade_id: str
    underlying_symbol: str
    rank: int
    strategy: str
    entry_mode: str | None = None
    scenario_tag: str
    primary_signal: str
    score: float
    confidence: float
    recommendation_snapshot: dict[str, Any]
    opened_at: datetime
    config_snapshot_id: str = "defaults"


class TradeOutcomeRecord(BaseModel):
    """Closed trade with full lineage for attribution (§12.7)."""

    outcome_id: str
    trade_id: str
    underlying_symbol: str
    rank: int
    strategy: str
    entry_mode: str | None = None
    scenario_tag: str
    primary_signal: str
    score_at_entry: float
    confidence_at_entry: float
    outcome: TradeOutcomeLabel
    realized_pnl_inr: float
    exit_reason: str
    notes: str | None = None
    recommendation_snapshot: dict[str, Any]
    opened_at: datetime
    closed_at: datetime
    failure_memory_id: str | None = None
    config_snapshot_id: str = "defaults"


class FailureMemoryEntry(BaseModel):
    """Structured losing-trade context stored for pre-trade retrieval."""

    failure_id: str
    trade_id: str
    underlying_symbol: str
    strategy: str
    entry_mode: str | None = None
    scenario_tag: str
    primary_signal: str
    market_summary: str
    loss_pnl_inr: float
    exit_reason: str
    lesson: str
    tags: list[str] = Field(default_factory=list)
    closed_at: datetime


class ModuleAttribution(BaseModel):
    strategy: str
    trade_count: int
    wins: int
    losses: int
    scratches: int
    win_rate: float
    total_pnl_inr: float
    expectancy_inr: float
    weight: float = 1.0


class AdaptationEvent(BaseModel):
    adaptation_id: str
    triggered_at: datetime
    trigger: str
    action: str
    parameter_changes: dict[str, Any] = Field(default_factory=dict)
    walk_forward_passed: bool
    deployed: bool
    detail: str
    config_snapshot_id: str


class CloseTradeRequest(BaseModel):
    trade_id: str
    outcome: TradeOutcomeLabel
    realized_pnl_inr: float
    exit_reason: str
    notes: str | None = None


class LearningDashboard(BaseModel):
    """Operator view of the continual learning loop."""

    as_of: datetime
    closed_trade_count: int
    open_trade_count: int
    win_rate: float | None
    total_pnl_inr: float
    profit_factor: float | None
    failure_memory_count: int
    adaptation_frozen: bool
    freeze_reason: str | None = None
    module_attribution: list[ModuleAttribution]
    recent_outcomes: list[TradeOutcomeRecord]
    open_trades: list[OpenTradeRecord]
    failure_memories: list[FailureMemoryEntry]
    adaptation_history: list[AdaptationEvent]
    learning_loop_notes: list[str]

File: frontend/src/types/learning.ts
/** Continual learning types — architecture §12 */

export type TradeOutcomeLabel = "win" | "loss" | "scratch";

export interface FailureMemoryMatch {
  failure_id: string;
  similarity: number;
  summary: string;
  loss_pnl_inr: number;
  lesson: string;
  strategy: string;
  underlying_symbol: string;
  closed_at: string;
}

export interface LearningInsight {
  failure_matches: FailureMemoryMatch[];
  confidence_penalty: number;
  confidence_before: number;
  confidence_after: number;
  module_win_rate?: number | null;
  module_trade_count: number;
  module_expectancy_inr?: number | null;
  learning_note: string;
}

export interface OpenTradeRecord {
  trade_id: string;
  underlying_symbol: string;
  rank: number;
  strategy: string;
  entry_mode?: string | null;
  scenario_tag: string;
  primary_signal: string;
  score: number;
  confidence: number;
  opened_at: string;
  config_snapshot_id: string;
}

export interface TradeOutcomeRecord {
  outcome_id: string;
  trade_id: string;
  underlying_symbol: string;
  rank: number;
  strategy: string;
  entry_mode?: string | null;
  scenario_tag: string;
  primary_signal: string;
  score_at_entry: number;
  confidence_at_entry: number;
  outcome: TradeOutcomeLabel;
  realized_pnl_inr: number;
  exit_reason: string;
  notes?: string | null;
  opened_at: string;
  closed_at: string;
  failure_memory_id?: string | null;
  config_snapshot_id: string;
}

export interface FailureMemoryEntry {
  failure_id: string;
  trade_id: string;
  underlying_symbol: string;
  strategy: string;
  entry_mode?: string | null;
  scenario_tag: string;
  primary_signal: string;
  market_summary: string;
  loss_pnl_inr: number;
  exit_reason: string;
  lesson: string;
  tags: string[];
  closed_at: string;
}

export interface ModuleAttribution {
  strategy: string;
  trade_count: number;
  wins: number;
  losses: number;
  scratches: number;
  win_rate: number;
  total_pnl_inr: number;
  expectancy_inr: number;
  weight: number;
}

export interface AdaptationEvent {
  adaptation_id: string;
  triggered_at: string;
  trigger: string;
  action: string;
  parameter_changes: Record<string, unknown>;
  walk_forward_passed: boolean;
  deployed: boolean;
  detail: string;
  config_snapshot_id: string;
}

export interface LearningDashboard {
  as_of: string;
  closed_trade_count: number;
  open_trade_count: number;
  win_rate: number | null;
  total_pnl_inr: number;
  profit_factor: number | null;
  failure_memory_count: number;
  adaptation_frozen: boolean;
  freeze_reason?: string | null;
  module_attribution: ModuleAttribution[];
  recent_outcomes: TradeOutcomeRecord[];
  open_trades: OpenTradeRecord[];
  failure_memories: FailureMemoryEntry[];
  adaptation_history: AdaptationEvent[];
  learning_loop_notes: string[];
}

export interface CloseTradeRequest {
  trade_id: string;
  outcome: TradeOutcomeLabel;
  realized_pnl_inr: number;
  exit_reason: string;
  notes?: string;
}

export const STRATEGY_LABELS: Record<string, string> = {
  simple_volatility: "Simple Volatility",
  gamma_scalping: "Gamma Scalping",
  vega_scalping: "Vega Scalping",
};
